### PR TITLE
IDSEQ-1759 Check reference cache against S3 object etag

### PR DESCRIPTION
**Pull request by kislyuk**
This guarantees that the local file is byte-identical to the one on S3.

Tested on a test instance running idseq-dag.

### CODE REVIEW

File: idseq_dag/util/s3.py [hunk 1]
```diff
@@ -140,9 +152,22 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
...
         if os.path.exists(dst):
-            # No need to fetch this file from s3, it has been just produced
-            # on this instance.
-            return dst
+            # Destination filename exists. If it is in the reference download directory, check the reference fetch log
+            # for a record of the source key and etag, and short-circuit the download if they match. Otherwise, blindly
+            # assume that it's the same file and short-circuit the download.
+            if abspath.startswith(config["REF_DIR"]):
+                try:
+                    with open(os.path.join(config["REF_FETCH_LOG_DIR"], abspath_hash)) as fh:
```
Comment (davidrissato) [inline idseq_dag/util/s3.py]: Considering IDseq uses `multiprocessing` package and have tons of separate processes, should we protect the code section that is reading / writing json files with a multiprocessing Lock?

File: idseq_dag/util/s3.py [hunk 2]
```diff
@@ -203,6 +228,11 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
                             named_args=named_args
                         )
                     )
+                if abspath.startswith(config["REF_DIR"]):
+                    os.makedirs(config["REF_FETCH_LOG_DIR"], exist_ok=True)
+                    with open(os.path.join(config["REF_FETCH_LOG_DIR"], abspath_hash), "w") as fh:
```
Comment (davidrissato) [inline idseq_dag/util/s3.py]: We may want to protect this section with a multiprocessing lock as well

File: idseq_dag/util/s3.py (hunk 1 above)
Comment (kislyuk) [reply idseq_dag/util/s3.py]: No, because this section is protected by the lock on line 139/151 (https://github.com/chanzuckerberg/idseq-dag/pull/218/files#diff-30956c749673657f32d6e09a0ac2e49dR151).

File: idseq_dag/util/s3.py (hunk 2 above)
Comment (kislyuk) [reply idseq_dag/util/s3.py]: No, this line is also protected by the lock on line 139/151 (https://github.com/chanzuckerberg/idseq-dag/pull/218/files#diff-30956c749673657f32d6e09a0ac2e49dR151).

I am also puzzled by this design pattern of putting locks everywhere. Generally speaking, that's what leads to deadlocks. It is preferable to structure operations in such a way that they either share nothing, or move data into place atomically or over such a short period of time that a race is implausible. That way, even if they end up racing against each other, the processes will produce the same result.

Comment (davidrissato) [reply idseq_dag/util/s3.py]: The current design heavily relies on multiprocessing, semaphore and locks, which makes it hard to follow the logic. I hadn't noticed those locks you mentioned and I agree they should be enough to protect this code section.

Comment (davidrissato) [reply idseq_dag/util/s3.py]: Hmm, I'm reviewing it again and I believe those locks may not be enough. They are related to a specific destination file. Therefore, if two different processes are downloading different files, this code section will be unprotected from concurrency issues.

Comment (kislyuk) [reply idseq_dag/util/s3.py]: What are the concurrency issues that you're worried about?

Comment (davidrissato) [reply idseq_dag/util/s3.py]: I just realized you are creating a json file for each destination file. You are right, that lock should prevent any concurrency issues.

**Comment by boris-dimitrov** [comment]
I would be delighted to explain this design or review it with you.  Locks do not lead to deadlock if acquired in a partial order hierarchy.  The ones I introduced here are provably deadlock free.

I am however concerned by the use of boto3 here considering this code will run from multiple threads and processes and may share global boto3 session state.  According to boto documentation that is not advised.  It is one reason we refrained from using boto in this codebase until evidently a few month ago.  Also coincidentally we have noticed some deadlocks.

File: idseq_dag/util/s3.py [hunk 3]
```diff
@@ -26,13 +29,20 @@
 S3MI_SEM = multiprocessing.Semaphore(MAX_CONCURRENT_S3MI_DOWNLOADS)
 IOSTREAM_UPLOADS = multiprocessing.Semaphore(MAX_CONCURRENT_UPLOAD_OPERATIONS)
 
+s3 = boto3.resource('s3')
```
Comment (boris-dimitrov) [inline idseq_dag/util/s3.py]: Is this safe to use from multiple threads and from multiple processes?

File: idseq_dag/util/s3.py [hunk 4]
```diff
@@ -140,9 +152,22 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
...
-            # on this instance.
-            return dst
+            # Destination filename exists. If it is in the reference download directory, check the reference fetch log
+            # for a record of the source key and etag, and short-circuit the download if they match. Otherwise, blindly
+            # assume that it's the same file and short-circuit the download.
+            if abspath.startswith(config["REF_DIR"]):
+                try:
+                    with open(os.path.join(config["REF_FETCH_LOG_DIR"], abspath_hash)) as fh:
+                        fetch_record = json.load(fh)
+                    obj = s3.Bucket(parsed_s3_url.netloc).Object(parsed_s3_url.path.lstrip('/'))
```
Comment (boris-dimitrov) [inline idseq_dag/util/s3.py]: Does this fail if the object doesn't exist in s3?  Or is it just hashing just the URL?

I ask because in parallel with you I was trying to cleanup and reduce calls to s3 from the wrapper function around this whole function, which tries to guess the correct compression extension and checks repeatedly if various compressed versions of the file exist.   I wanted to make sure that check only happens if the file actually needs to be downloaded. m

File: idseq_dag/util/s3.py [hunk 5]
```diff
@@ -140,9 +152,22 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
...
+                    assert fetch_record["bucket_name"] == obj.bucket_name
+                    assert fetch_record["key"] == obj.key
+                    assert fetch_record["e_tag"] == obj.e_tag
+                    return dst
+                except Exception:
+                    pass
```
Comment (boris-dimitrov) [inline idseq_dag/util/s3.py]: Shouldn't this "rm -rf {dst}" instead of "pass"?

It's a minor concern, but the fall-through path previously only ran when "dst" didn't exist, so it may be better to ensure that.

In practice this can be noticed only when automatically untarrig "foo.tar.gz" into "foo/..." and without the deleting we would see some files from the previous download even if the file is downloaded and untarred again on top.

File: idseq_dag/util/s3.py (hunk 2 above)
Comment (boris-dimitrov) [reply idseq_dag/util/s3.py]: We mostly rely on multithreading for concurrency.   We hardly ever use multiprocessing, and in such a limited way that it would be rather straightforward to replace with actual subcommands.  We literally never communicate back from child to parent, except through a file.  We use multiprocessing locks because they are safe for both multithreading and multiprocessing.  Unfortunately, the same cannot be said about some library modules, particularly the python logging package, which has often been a source of issues with multiprocessing.  For this reason, I would be quite open to removing the use of multiprocess start/end and replacing with subcommands.  Shouldn't take more than a day or two.

File: idseq_dag/util/s3.py (hunk 4 above)
Comment (boris-dimitrov) [reply idseq_dag/util/s3.py]: I think if I understand this design correctly, it definitely polls S3, which is good for ensuring that it always fetches the newest file (if it got updated).   I think this is more valuable than trying to reduce S3 calls.  Thanks.

Comment (boris-dimitrov) [reply idseq_dag/util/s3.py]: Back to David's concern about locking.   I think, unfortunately, locks[abspath] aka destination_lock is still a threading.RLock, from the days before we used multiprocessing.  It should be a multiprocessing.RLock now for extra safety.  Same with mutex.  

I'll also restate my question - is it safe to use the same global boto3 session from multiple threads and from multiple processes?   If the answer is no, it would rather easy to solve the problem here by creating a separate session for each destination.  The multiple threads/processes serialize by destination --- this is what the locks in this design ensure.  A session per call would be fine too, for fetch_from_s3.  I am more concerned about that with check_s3_presence.

File: idseq_dag/util/s3.py (hunk 5 above)
Comment (kislyuk) [reply idseq_dag/util/s3.py]: I consider it better/more minimalistic to let the following command overwrite the file.

File: idseq_dag/util/s3.py (hunk 4 above)
Comment (kislyuk) [reply idseq_dag/util/s3.py]: There is some uncertainty about how safe it is to use the same boto3 client instance from multiple threads, so I'm going to make each subroutine spawn its own client.

File: idseq_dag/util/s3.py (hunk 3 above)
Comment (kislyuk) [reply idseq_dag/util/s3.py]: There is some uncertainty about how safe it is to use the same boto3 client instance from multiple threads, so I'm going to make each subroutine spawn its own client.

**Comment by kislyuk** [comment]
Pushed https://github.com/chanzuckerberg/idseq-dag/commit/6abd9ab to address doubts about the s3 thread safety.

File: idseq_dag/util/s3.py (hunk 2 above)
Comment (kislyuk) [reply idseq_dag/util/s3.py]: In Python, I prefer to use concurrent.futures.ThreadPoolExecutor for concurrency, and nothing else. That guides developers into a particular programming model that is far safer than dealing with locks individually. While I have no doubt that you can write threadsafe code using basic locking primitives, I don't trust myself to do so, and I don't think developers in general will be able to safely change your code.

With that said, I don't think we should be improving this codebase any further; instead we should focus on replacing it with the other alternatives we discussed (like miniwdl) that share nothing and use concurrency on a lower level. This discussion only strengthens and underscores my conviction that we should avoid investing in this codebase and retire it as quickly as possible, loading and running code in the individual steps in https://github.com/chanzuckerberg/idseq-dag/tree/master/idseq_dag/steps from an external executor as a migration strategy.

File: idseq_dag/util/s3.py (hunk 5 above)
Comment (boris-dimitrov) [reply idseq_dag/util/s3.py]: It may be not just a single file but a subdirectory untarred.   In which case it wouldn't delete any files left over that shouldn't be present in the new version.

File: idseq_dag/util/s3.py (hunk 4 above)
Comment (boris-dimitrov) [reply idseq_dag/util/s3.py]: Thank you.

File: idseq_dag/util/s3.py (hunk 3 above)
Comment (boris-dimitrov) [reply idseq_dag/util/s3.py]: Thank you.

File: idseq_dag/util/s3.py (hunk 2 above)
Comment (boris-dimitrov) [reply idseq_dag/util/s3.py]: I am open to using the pattern you recommend, but unclear on how it relates to the pattern when multiple of these threads wish to download the same file, and something in your infrastructure ought to ensure the file is downloaded just once.   Clearly that requires some kind of locking, if threads are used.

Separately, and more importantly, even if our plan is to completely deprecate this infrastructure and switch to something better, I think you would agree with me that in the meantime we should at least not make it less stable.   I see you agree because you decided to create private boto sessions for every thread.  That's wonderful.
